### PR TITLE
fix: hrmp

### PR DESCRIPTION
**Pull request by evilrobot-01**
Ensures hrmp channels work seamlessly via `pop-cli`.

Requires v1.3.3+ of the Paseo chain spec generator.

Test using:
`cargo run -- up parachain -f ./tests/networks/pop.toml --verbose`

### CODE REVIEW

**Comment by Daanvdplas** [comment]
I noticed that when I build the chain-spec-generator and used it to spin up a local network the parachains were not producing blocks. Used the release binary and it worked. Not 100% sure I did something wrong or it is the chain spec generator. Just put it here so that we are aware.

**Comment by evilrobot-01** [comment]
> Makes sense to replace https://github.com/r0gue-io/paseo-runtimes with https://github.com/r0gue-io/paseo-network-runtimes and begin maintaining it.

Need to coordinate with Ale in terms of whether this could be upstreamed to Paseo repo, even if with some feature flag.

**Comment by evilrobot-01** [comment]
> As for the hrmp at genesis that can and should be pushed upstream! There's no point on us keeping a fork if there's not that many changes.
> 
> I can take care of doing so 🤝

Should just be whatever is applicable at https://github.com/paseo-network/runtimes/compare/main...r0gue-io:paseo-network-runtimes:pop. It would be great to get this into the next pop release, but that would require a Paseo release with the updated CSG.

**Comment by al3mart** [comment]
@evilrobot-01 v1.3.3 and v1.3.4 should present no problems now.

**Comment by evilrobot-01** [comment]
Rebased

File: tests/networks/pop.toml [hunk 1]
```diff
@@ -24,4 +24,16 @@ default_command = "pop-node"
 
 [[parachains.collators]]
 name = "pop"
-args = ["-lruntime::contracts=debug"]
\ No newline at end of file
+args = ["-lruntime::contracts=debug"]
+
+[[hrmp_channels]]
```
Comment (evilrobot-01) [inline tests/networks/pop.toml:29]: Has channels for testing. Without this PR, this network config should result in no parachain blocks produced.

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 2]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
 				}
 
 				spinner.stop(result);
+
+				// Check for any specified channels
```
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/up/parachain.rs:151]: No CLI tests sadly, suggest they are tackled as a separate PR.

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 3]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
...
+				if zombienet.hrmp_channels() {
+					let relay_chain = zombienet.relay_chain();
+					match RelayChain::from(relay_chain) {
+						None => {
+							log::error(format!("🚫 Using `{relay_chain}` with HRMP channels is currently unsupported. Please use `paseo-local` or `rococo-local`."))?;
+						},
+						Some(_) => {
+							use tokio::time::sleep;
```
Comment (peterwht) [inline crates/pop-cli/src/commands/up/parachain.rs:159]:
```suggestion
```

Is imported already

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 4]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
...
+							let spinner = cliclack::spinner();
```
Comment (peterwht) [inline crates/pop-cli/src/commands/up/parachain.rs:160]:
```suggestion
							let spinner = spinner();
```

Any reason not to do a top level import of spinner?

File: crates/pop-parachains/src/relay.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	config::BlockHash,
+	dynamic::{self, Value},
+	ext::sp_core,
+	OnlineClient, PolkadotConfig,
+};
+
+/// Clears the DMPQ state for the given parachain IDs.
+///
+/// # Arguments
+/// * `relay_chain` - The relay chain.
```
Comment (peterwht) [inline crates/pop-parachains/src/relay.rs:15]:
```suggestion
```
this argument doesn't exist anymore

File: crates/pop-parachains/src/up/mod.rs [hunk 6]
```diff
@@ -216,7 +220,7 @@ impl Zombienet {
 	///   will use the latest available version).
 	/// * `network_config` - The network configuration to be used to launch a network.
 	/// * `cache` - The location used for caching binaries.
-	async fn relay_chain(
+	async fn _relay_chain(
```
Comment (peterwht) [inline crates/pop-parachains/src/up/mod.rs:223]: Why the `_` prefix? This infers the function is not used.

If it's to not conflict with the other `relay_chain`, maybe something like `relay_chain_config` would be more accurate?

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 7]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
...
+							spinner.start("Connecting to relay chain to prepare channels...");
+							// Allow relay node time to start
+							sleep(Duration::from_secs(10)).await;
+							spinner.start("Preparing channels...");
```
Comment (peterwht) [inline crates/pop-cli/src/commands/up/parachain.rs:164]: I haven't ran this code, but I assume that this second `spinner.start` simply replaces the original spinner message (3 lines above).

In this case, won't `spinner.set_message("...")` be more clear?

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 8]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
...
+							let relay_endpoint = network.relaychain().nodes()[0].client().await?;
+							let para_ids: Vec<_> =
+								network.parachains().iter().map(|p| p.para_id()).collect();
+							tokio::spawn(async move {
+								if let Err(e) = clear_dmpq(relay_endpoint, &para_ids).await {
+									spinner.stop("");
+									log::error(format!("🚫 Could not prepare channels: {e}"))?;
```
Comment (peterwht) [inline crates/pop-cli/src/commands/up/parachain.rs:171]:
```suggestion
									spinner.error(format!("🚫 Could not prepare channels: {e}"));
```

File: crates/pop-parachains/src/relay.rs (hunk 5 above)
Comment (evilrobot-01) [reply crates/pop-parachains/src/relay.rs:15]: Sloppy, apologies.

File: crates/pop-parachains/src/up/mod.rs (hunk 6 above)
Comment (evilrobot-01) [reply crates/pop-parachains/src/up/mod.rs:223]: _ is a valid function name, its just linting will not complain. Will update.

File: crates/pop-cli/src/commands/up/parachain.rs [hunk 9]
```diff
@@ -147,6 +147,37 @@ impl ZombienetCommand {
...
 
 				spinner.stop(result);
+
+				// Check for any specified channels
+				if zombienet.hrmp_channels() {
+					let relay_chain = zombienet.relay_chain();
+					match RelayChain::from(relay_chain) {
+						None => {
+							log::error(format!("🚫 Using `{relay_chain}` with HRMP channels is currently unsupported. Please use `paseo-local` or `rococo-local`."))?;
+						},
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply crates/pop-cli/src/commands/up/parachain.rs:157]: Fixed in bc8f821d

File: crates/pop-parachains/src/up/mod.rs (hunk 6 above)
Comment (evilrobot-01) [reply crates/pop-parachains/src/up/mod.rs:223]: Improved in 3b3774a9

File: crates/pop-cli/src/commands/up/parachain.rs (hunk 7 above)
Comment (evilrobot-01) [reply crates/pop-cli/src/commands/up/parachain.rs:163]: Improved in 2e71e04b

File: crates/pop-cli/src/commands/up/parachain.rs (hunk 4 above)
Comment (evilrobot-01) [reply crates/pop-cli/src/commands/up/parachain.rs]: Name clash with import. Have changed to `progress` in 2e71e04b

File: crates/pop-cli/src/commands/up/parachain.rs (hunk 8 above)
Comment (evilrobot-01) [reply crates/pop-cli/src/commands/up/parachain.rs:170]: Implemented in 23ec1a68

**Review by AlexD10S — APPROVED** [review]
Looks great to me, and I see you’ve addressed all of Peter’s comments. 
Thank you for the hard work you’ve put into this!

**Comment by evilrobot-01** [comment]
> Looks great to me, and I see you’ve addressed all of Peter’s comments.

Assume they will need resolving before merging. Figure its best to have whomever is merging to resolve to ensure they are happy, rather than me resolving and things get hidden.

File: crates/pop-parachains/src/utils/helpers.rs [hunk 10]
```diff
@@ -87,6 +88,7 @@ mod tests {
 	use super::*;
 	use crate::generator::parachain::ChainSpec;
 	use askama::Template;
+	use std::env::var;
```
Comment (al3mart) [inline crates/pop-parachains/src/utils/helpers.rs:91]: Unused import

**Pull request merged**